Sources/AsyncHTTPClient/AsyncAwait/Transaction+StateMachine.swift

```diff
@@ -0,0 +1,603 @@
+
+            case .executing(let context, _, .finished(let streamID, let continuation)):
+                self.state = .finished(error: error, streamID)
+                return .failResponseStream(continuation, error, context.executor)
+            }
+        }
+
+        // MARK: - Request -
+
+        enum StartExecutionAction {
+            case cancel(HTTPRequestExecutor)
+            case none
+        }
+
+        mutating func willExecuteRequest(_ executor: HTTPRequestExecutor) -> StartExecutionAction {
+            switch self.state {
+            case .initialized(let continuation), .queued(let continuation, _):
+                let context = ExecutionContext(
+                    executor: executor,
+                    allocator: .init(),
+                    continuation: continuation
+                )
+                self.state = .executing(context, .initialized, .initialized)
+                return .none
+
+            case .finished(error: .some, .none):
+                return .cancel(executor)
+
+            case .executing,
+                 .finished(error: .none, _),
+                 .finished(error: .some, .some):
+                preconditionFailure("Invalid state: \(self.state)")
+            }
+        }
+
+        enum ResumeProducingAction {
+            case resumeStream(ByteBufferAllocator)
+            case none
+        }
+
+        mutating func resumeRequestBodyStream() -> ResumeProducingAction {
+            switch self.state {
+            case .initialized, .queued:
+                preconditionFailure("Received a resumeBodyRequest on a request, that isn't executing. Invalid state: \(self.state)")
+
+            case .executing(let context, .initialized, .initialized):
+                // the request can start to send its body.
+                self.state = .executing(context, .producing, .initialized)
+                return .resumeStream(context.allocator)
+
+            case .executing(_, .producing, _):
+                preconditionFailure("Received a resumeBodyRequest on a request, that is producing. Invalid state: \(self.state)")
+
+            case .executing(let context, .paused, let responseState):
+                // the request body was paused. we can start the body streaming again.
+                self.state = .executing(context, .producing, responseState)
+                return .resumeStream(context.allocator)
+
+            case .executing(_, .finished, _):
+                // the channels writability changed to writable after we have forwarded all the
+                // request bytes. Can be ignored.
+                return .none
+
+            case .executing(_, .initialized, .waitingForStream),
+                 .executing(_, .initialized, .buffering),
+                 .executing(_, .initialized, .waitingForRemote),
+                 .executing(_, .initialized, .finished):
+                preconditionFailure("Invalid state: Response can not be received before request. \(self.state)")
```

Comment:
This seems odd: haven't we sent the request head now? We haven't sent the request _body_, but that's a different question. Are we confident there's no interleaving of events that allows this to be hit?

Reply:
Changed the name of `initialized` to `requestHeadSent` and on the response side I changed `initialized` to `waitingForResponseHead`. This made my error here very obvious!